### ISSUE TITLE
Testing TauP-3.0.0 Pre-release 

### ISSUE THREAD

**Issue by luca-s** [crotwell/TauP#34]
Dear @crotwell ,

I gave a try to the pre-release TauP-3.0.0 and I noticed the addition of two interesting options to the `taup table` command: `--stadepth`,`--evdepth,` . Could you briefly explain how they affect the program behaviour? I can guess it but I would prefer to be sure about it.

I also have a very small note on the `--csv` option. Would it be possible to increase the precision of the columns (at least `Distance` and  `Depth`)? 

Thank you for your work, it looks like the next release will have many interesting additions.

Luca

**Comment by crotwell** [comment]
`--evdepth` is the same as -h, so source depth. 
`--stadepth` is the receiver depth if not at the surface

You can see the (simple) explanation with the cmd line help, like:
`taup table --help`

You can configre the precision of the outputs via a prop file. See  the`--prop <propfile>` parameter. For distance and depth you would change taup.depth.precision and  taup.distance.precision to be however many decimal digits you want. The defaults look like:
```

# precision for depth output, default is 1 decimal digits
# note that this is precision, not accuracy. Just because you get more
# digits doesn't imply that they have any meaning
taup.depth.precision=1

# precision for distance output, default is 2 decimal digits.
# note that this is precision, not accuracy. Just because you get more
# digits doesn't imply that they have any meaning
taup.distance.precision=2

```

**Comment by luca-s** [comment]
Thank you for your reply. The `propfile` parameter is great, I didn't know about it. 

I also read the help, but I am still confused about `--stadepth` ans `--evdepth` options. I would like to show you an example that might help clarify things.

This is the header used in the examples:
```
$ cat my.header 
n # P,S very dense travel-time table sampling for mymodel
3     # number of depth samples
   0.00   5.00   20.00
3    # number of distances
   0.00   0.01   0.20
```

Here are the results of `taup table` without neither `--stadepth` nor `--evdepth` option.

```
$ taup table  --mod iasp91 --ph ttp --header my.header  --csv -o out
$ cat out.csv 
Model,Distance (deg),Depth (km),Phase,Time (s),RayParam (deg/s),Takeoff Angle,Incident Angle,Purist Distance,Purist Name
iasp91,0.00,0.0,P,0.00,19.172,90.00,90.00,0.00,P
iasp91,0.00,0.0,PKiKP,994.57,0.000,0.00,0.00,0.00,PKikp
iasp91,0.01,0.0,P,0.19,19.172,89.99,89.99,0.01,P
iasp91,0.01,0.0,PKiKP,994.57,0.000,0.00,0.00,0.01,PKikp
iasp91,0.20,0.0,P,3.83,19.172,89.90,89.90,0.20,P
iasp91,0.20,0.0,PKiKP,994.57,0.004,0.01,0.01,0.20,PKikp
iasp91,0.00,5.0,P,0.00,19.172,90.00,90.00,0.00,P
iasp91,0.00,5.0,PKiKP,994.57,0.000,0.00,0.00,0.00,PKikp
iasp91,0.01,5.0,P,0.19,19.172,89.99,89.99,0.01,P
iasp91,0.01,5.0,PKiKP,994.57,0.000,0.00,0.00,0.01,PKikp
iasp91,0.20,5.0,P,3.83,19.172,89.90,89.90,0.20,P
iasp91,0.20,5.0,PKiKP,994.57,0.004,0.01,0.01,0.20,PKikp
iasp91,0.00,20.0,P,0.00,19.172,90.00,90.00,0.00,P
iasp91,0.00,20.0,PKiKP,994.57,0.000,0.00,0.00,0.00,PKikp
iasp91,0.01,20.0,P,0.19,19.172,89.99,89.99,0.01,P
iasp91,0.01,20.0,PKiKP,994.57,0.000,0.00,0.00,0.01,PKikp
iasp91,0.20,20.0,P,3.83,19.172,89.90,89.90,0.20,P
iasp91,0.20,20.0,PKiKP,994.57,0.004,0.01,0.01,0.20,PKikp
```

Here are the results with `stadepth` = 0.1. My understanding is that I can use the option to alter all the depths provided in `my.header` and make them as if they were `stadepth` km lower.  However it doesn't seem so, because a variation of 0.1 km shouldn't make that of a difference, but the results show differently.

```
$ taup table  --mod iasp91 --ph ttp --stadepth 0.1 --header my.header  --csv -o out
$ cat out.csv 
Model,Distance (deg),Depth (km),Phase,Time (s),RayParam (deg/s),Takeoff Angle,Incident Angle,Purist Distance,Purist Name
iasp91,0.00,0.0,PKiKP,994.55,0.000,0.00,0.00,0.00,PKikp
iasp91,0.01,0.0,PKiKP,994.55,0.000,0.00,0.00,0.01,PKikp
iasp91,0.20,0.0,PKiKP,994.55,0.004,0.01,0.01,0.20,PKikp
iasp91,0.00,5.0,PKiKP,994.55,0.000,0.00,0.00,0.00,PKikp
iasp91,0.01,5.0,PKiKP,994.55,0.000,0.00,0.00,0.01,PKikp
iasp91,0.20,5.0,PKiKP,994.55,0.004,0.01,0.01,0.20,PKikp
iasp91,0.00,20.0,PKiKP,994.55,0.000,0.00,0.00,0.00,PKikp
iasp91,0.01,20.0,PKiKP,994.55,0.000,0.00,0.00,0.01,PKikp
iasp91,0.20,20.0,PKiKP,994.55,0.004,0.01,0.01,0.20,PKikp
```

Finally here are the results with `evdepth` = 0.1, but I am still confused abut it. In which way is it supposed to alter the travel time table? Shouldn't be the header the one defining the depths to copute the travel times  for?

```
$ taup table  --mod iasp91 --ph ttp --stadepth 0 --header my.header --evdepth 0.1 --csv -o out
$ cat out.csv 
Model,Distance (deg),Depth (km),Phase,Time (s),RayParam (deg/s),Takeoff Angle,Incident Angle,Purist Distance,Purist Name
iasp91,0.00,0.0,p,0.02,0.000,180.00,0.00,0.00,p
iasp91,0.00,0.0,PKiKP,994.55,0.000,0.00,0.00,0.00,PKikp
iasp91,0.01,0.0,p,0.19,19.094,95.13,84.86,0.01,p
iasp91,0.01,0.0,PKiKP,994.55,0.000,0.00,0.00,0.01,PKikp
iasp91,0.20,0.0,p,3.83,19.171,90.21,89.62,0.20,p
iasp91,0.20,0.0,PKiKP,994.55,0.004,0.01,0.01,0.20,PKikp
iasp91,0.00,5.0,p,0.02,0.000,180.00,0.00,0.00,p
iasp91,0.00,5.0,PKiKP,994.55,0.000,0.00,0.00,0.00,PKikp
iasp91,0.01,5.0,p,0.19,19.094,95.13,84.86,0.01,p
iasp91,0.01,5.0,PKiKP,994.55,0.000,0.00,0.00,0.01,PKikp
iasp91,0.20,5.0,p,3.83,19.171,90.21,89.62,0.20,p
iasp91,0.20,5.0,PKiKP,994.55,0.004,0.01,0.01,0.20,PKikp
iasp91,0.00,20.0,p,0.02,0.000,180.00,0.00,0.00,p
iasp91,0.00,20.0,PKiKP,994.55,0.000,0.00,0.00,0.00,PKikp
iasp91,0.01,20.0,p,0.19,19.094,95.13,84.86,0.01,p
iasp91,0.01,20.0,PKiKP,994.55,0.000,0.00,0.00,0.01,PKikp
iasp91,0.20,20.0,p,3.83,19.171,90.21,89.62,0.20,p
iasp91,0.20,20.0,PKiKP,994.55,0.004,0.01,0.01,0.20,PKikp
```


Thank you for your patience.

Luca

**Comment by crotwell** [comment]
Ah, yes, sorry was thinking about the parameters, but not in terms of taup table. So didn't actually answer the question you asked.

Short answer is it is a bug. The event depth argument, as you say, should not exist in table command as that list of depths comes from the input file, not a command line arg. Station depth should be allowed, and would modify the receiver depth for all calculations. 

The large difference in the output of table in your example is not travel time difference, but rather in whether or not a phase exists. So for a receiver at even a small depth like 0.1, the phase "P" doesn't exist as it cannot go down, turn and arrive at the station for very close distances. Using ttp can be very useful, but also can be confusing because if a phase doesn't exist, like P, taup will pick the first of whatever phases do arrive, like PKiKP, which may not be what you actually want. Unfortunately, because "ttp" comes from the old ttimes program, it does not include the Ped phase as we made it up much later and receiver at depth wasn't possible in that older code.

For example, limiting phases to just the three non-reflecting P paths, so p, Ped and P, a surface source and receiver has all P arrivals as first:
```
 bin/taup table -o stdout --csv --stadepth 0.0 -p p,Ped,P | head
Model,Distance (deg),Depth (km),Phase,Time (s),RayParam (deg/s),Takeoff Angle,Incident Angle,Purist Distance,Purist Name
iasp91,0.00,0.0,P,0.00,19.172,90.00,90.00,0.00,P
iasp91,0.10,0.0,P,1.92,19.172,89.95,89.95,0.10,P
iasp91,0.20,0.0,P,3.83,19.172,89.90,89.90,0.20,P
iasp91,0.30,0.0,P,5.75,19.171,89.85,89.85,0.30,P
iasp91,0.40,0.0,P,7.67,19.171,89.80,89.80,0.40,P
iasp91,0.50,0.0,P,9.59,19.171,89.75,89.75,0.50,P
iasp91,0.60,0.0,P,11.50,19.171,89.70,89.70,0.60,P
iasp91,0.70,0.0,P,13.42,19.171,89.65,89.65,0.70,P
iasp91,0.80,0.0,P,15.34,19.171,89.60,89.60,0.80,P
```

But if we move the receiver to a slight depth, the P is replaced by the "exclusively downgoing" P wave, Ped, that doesn't turn. Then at about 0.6 degrees we are far enough away from the source that P can turn and arrive.

```
bin/taup table -o stdout --csv --stadepth 0.1 -p p,Ped,P | head
Model,Distance (deg),Depth (km),Phase,Time (s),RayParam (deg/s),Takeoff Angle,Incident Angle,Purist Distance,Purist Name
iasp91,0.00,0.0,Ped,0.02,0.000,0.00,180.00,0.00,Ped
iasp91,0.10,0.0,Ped,1.92,19.170,89.27,90.65,0.10,Ped
iasp91,0.20,0.0,Ped,3.83,19.171,89.62,90.21,0.20,Ped
iasp91,0.30,0.0,Ped,5.75,19.171,89.68,90.02,0.30,Ped
iasp91,0.40,0.0,P,7.67,19.171,89.67,89.93,0.40,P
iasp91,0.50,0.0,P,9.59,19.171,89.65,89.85,0.50,P
iasp91,0.60,0.0,P,11.50,19.171,89.61,89.78,0.60,P
iasp91,0.70,0.0,P,13.42,19.171,89.58,89.72,0.70,P
iasp91,0.80,0.0,P,15.34,19.171,89.53,89.66,0.80,P
```

P and Ped are kind of the same, but a reversal of P and p for source at depth. The distances at which Ped exists for a surface source to a receiver at 0.1km depth is exactly the same as "p" existing for an event at 0.1km to a surface receiver, just the path is traversed backwards.

I would probably suggest using your own list of phases instead of ttp for taup table, just so you know exactly the phases you are asking for.

Let me know if this doesn't make sense...

**Comment by luca-s** [comment]
Thank you for taking the time to go through this clear explanation. I definitely misunderstood the phase names, but with your explanation it makes now sense. 

---

However I believe I found a bug on `taup_table` when `--csv` option is used. Please see the following example: `taup_time`  and `taup_table` agree on the phase names and they also makes sense, but `taup_table + csv option` gives different results.


Here is `taup_time`, all seem good.

```
$ taup  time --ph p,Ped,P --stadepth 0.0  --evdepth=0 --deg=0.20
$ taup  time --ph p,Ped,P --stadepth 0.0  --evdepth=0 --deg=0.01
$ taup  time --ph p,Ped,P --stadepth 0.0  --evdepth=0 --deg=0.00
$ taup  time --ph p,Ped,P --stadepth 0.0  --evdepth=2 --deg=0.20
$ taup  time --ph p,Ped,P --stadepth 0.0  --evdepth=2 --deg=0.01
$ taup  time --ph p,Ped,P --stadepth 0.0  --evdepth=2 --deg=0.00
$ taup  time --ph p,Ped,P --stadepth 0.0  --evdepth=10 --deg=0.20
$ taup  time --ph p,Ped,P --stadepth 0.0  --evdepth=10 --deg=0.01
$ taup  time --ph p,Ped,P --stadepth 0.0  --evdepth=10 --deg=0.00

Model: iasp91
Distance   Depth   Phase   Travel    Ray Param  Takeoff  Incident Station   Purist   Purist
  (deg)     (km)   Name    Time (s)  p (s/deg)   (deg)    (deg)    (km)     Distance  Name 
-------------------------------------------------------------------------------------------
    0.20     0.0   P          3.83    19.172     89.90    89.90      0.0     0.20   = P    
    0.01     0.0   P          0.19    19.172     89.99    89.99      0.0     0.01   = P    
    0.00     0.0   P          0.00    19.172     90.00    90.00      0.0     0.00   = P    
    0.20     2.0   p          3.85    19.091     95.04    84.76      0.0     0.20   = p    
    0.01     2.0   p          0.39     9.313    150.93    29.06      0.0     0.01   = p    
    0.00     2.0   p          0.34     0.000    180.00     0.00      0.0     0.00   = p    
    0.20    10.0   p          4.20    17.469    114.13    65.67      0.0     0.20   = p    
    0.01    10.0   p          1.73     2.115    173.66     6.33      0.0     0.01   = p    
    0.00    10.0   p          1.72     0.000    180.00     0.00      0.0     0.00   = p    
```

```
$ taup  time --ph p,Ped,P --stadepth 0.2  --evdepth=0 --deg=0.20
$ taup  time --ph p,Ped,P --stadepth 0.2  --evdepth=0 --deg=0.01
$ taup  time --ph p,Ped,P --stadepth 0.2  --evdepth=0 --deg=0.00
$ taup  time --ph p,Ped,P --stadepth 0.2  --evdepth=2 --deg=0.20
$ taup  time --ph p,Ped,P --stadepth 0.2  --evdepth=2 --deg=0.01
$ taup  time --ph p,Ped,P --stadepth 0.2  --evdepth=2 --deg=0.00
$ taup  time --ph p,Ped,P --stadepth 0.2  --evdepth=10 --deg=0.20
$ taup  time --ph p,Ped,P --stadepth 0.2  --evdepth=10 --deg=0.01
$ taup  time --ph p,Ped,P --stadepth 0.2  --evdepth=10 --deg=0.00


Model: iasp91
Distance   Depth   Phase   Travel    Ray Param  Takeoff  Incident Station   Purist   Purist
  (deg)     (km)   Name    Time (s)  p (s/deg)   (deg)    (deg)    (km)     Distance  Name 
-------------------------------------------------------------------------------------------
    0.20     0.0   Ped        3.83    19.170     89.32    90.51      0.2     0.20   = Ped  
    0.01     0.0   Ped        0.19    18.864     79.72   100.27      0.2     0.01   = Ped  
    0.00     0.0   Ped        0.03     0.000      0.00   180.00      0.2     0.00   = Ped  
    0.20     2.0   p          3.85    19.106     94.53    85.27      0.2     0.20   = p    
    0.01     2.0   p          0.36    10.073    148.29    31.70      0.2     0.01   = p    
    0.00     2.0   p          0.31     0.000    180.00     0.00      0.2     0.00   = p    
    0.20    10.0   p          4.19    17.527    113.70    66.10      0.2     0.20   = p    
    0.01    10.0   p          1.70     2.158    173.53     6.46      0.2     0.01   = p    
    0.00    10.0   p          1.69     0.000    180.00     0.00      0.2     0.00   = p    
```

Here is `taup_table`, all seem good and the results agree with `taup_time`.

```
$ cat my.header 
n # P,S very dense travel-time table sampling for mymodel
3     # number of depth samples
   0.00   2.00   10.00
3    # number of distances
   0.20   0.01   0.00
```

```
$ taup table -o stdout  --stadepth 0.0 -p p,Ped,P  --header my.header
iasp91     0.20      0.0 P      3.83    19.172       0.20  P
iasp91     0.01      0.0 P      0.19    19.172       0.01  P
iasp91     0.00      0.0 P      0.00    19.172       0.00  P
iasp91     0.20      2.0 p      3.85    19.091       0.20  p
iasp91     0.01      2.0 p      0.39     9.313       0.01  p
iasp91     0.00      2.0 p      0.34     0.000       0.00  p
iasp91     0.20     10.0 p      4.20    17.469       0.20  p
iasp91     0.01     10.0 p      1.73     2.115       0.01  p
iasp91     0.00     10.0 p      1.72     0.000       0.00  p
```

```
$ taup  table -o stdout  --stadepth 0.2 -p p,Ped,P  --header my.header
iasp91     0.20      0.0 Ped      3.83    19.170       0.20  Ped
iasp91     0.01      0.0 Ped      0.19    18.864       0.01  Ped
iasp91     0.00      0.0 Ped      0.03     0.000       0.00  Ped
iasp91     0.20      2.0 p      3.85    19.106       0.20  p
iasp91     0.01      2.0 p      0.36    10.073       0.01  p
iasp91     0.00      2.0 p      0.31     0.000       0.00  p
iasp91     0.20     10.0 p      4.19    17.527       0.20  p
iasp91     0.01     10.0 p      1.70     2.158       0.01  p
iasp91     0.00     10.0 p      1.69     0.000       0.00  p
```

**Here is the bug, `taup_table + csv`:**

```
$ taup  table -o stdout --csv --stadepth 0.0 -p p,Ped,P  --header my.header
Model,Distance (deg),Depth (km),Phase,Time (s),RayParam (deg/s),Takeoff Angle,Incident Angle,Purist Distance,Purist Name
iasp91,0.20,0.0,P,3.83,19.172,89.90,89.90,0.20,P
iasp91,0.01,0.0,P,0.19,19.172,89.99,89.99,0.01,P
iasp91,0.00,0.0,P,0.00,19.172,90.00,90.00,0.00,P
iasp91,0.20,2.0,P,3.83,19.172,89.90,89.90,0.20,P
iasp91,0.01,2.0,P,0.19,19.172,89.99,89.99,0.01,P
iasp91,0.00,2.0,P,0.00,19.172,90.00,90.00,0.00,P
iasp91,0.20,10.0,P,3.83,19.172,89.90,89.90,0.20,P
iasp91,0.01,10.0,P,0.19,19.172,89.99,89.99,0.01,P
iasp91,0.00,10.0,P,0.00,19.172,90.00,90.00,0.00,P
```

```
$ taup  table -o stdout --csv --stadepth 0.2 -p p,Ped,P  --header my.header
Model,Distance (deg),Depth (km),Phase,Time (s),RayParam (deg/s),Takeoff Angle,Incident Angle,Purist Distance,Purist Name
iasp91,0.20,0.0,Ped,3.83,19.170,89.32,90.51,0.20,Ped
iasp91,0.01,0.0,Ped,0.19,18.864,79.72,100.27,0.01,Ped
iasp91,0.00,0.0,Ped,0.03,0.000,0.00,180.00,0.00,Ped
iasp91,0.20,2.0,Ped,3.83,19.170,89.32,90.51,0.20,Ped
iasp91,0.01,2.0,Ped,0.19,18.864,79.72,100.27,0.01,Ped
iasp91,0.00,2.0,Ped,0.03,0.000,0.00,180.00,0.00,Ped
iasp91,0.20,10.0,Ped,3.83,19.170,89.32,90.51,0.20,Ped
iasp91,0.01,10.0,Ped,0.19,18.864,79.72,100.27,0.01,Ped
iasp91,0.00,10.0,Ped,0.03,0.000,0.00,180.00,0.00,Ped
```

**Comment by crotwell** [comment]
believe I have this fixed in b06538c0 

Hope to have a new prerelease out soon, but have couple of other bugs to squash first.

thanks again for bug report, appreciate you taking the time to test.

**Comment by crotwell** [comment]
New prerelease if you are interested...
https://www.seis.sc.edu/downloads/TauP/prerelease/TauP-3.0.0-SNAPSHOT7.tgz
